docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,317 @@
+SP #004: Initialization
+=================
+
+This proposal documents the desired behavior of initialization related language semantics, including default constructor, initialization list and variable initialization.
+
+Status
+------
+
+Status: Design Review
+
+Implemtation: N/A
+
+Author: Yong He
+
+Reviewer: Theresa Foley, Kai Zhang
+
+Background
+----------
+
+Slang has introduced several different syntax around initialization to provide syntactic compatibility with HLSL/C++. As the language evolve, there are many corners where
+the semantics around initialization are not well-defined, and causing confusion or leading to surprising behaviors.
+
+This proposal attempts to provide a design on where we want the language to be in terms of how initialization is handled in all different places.
+
+Related Work
+------------
+
+C++ has many different ways and syntax to initialize an object: through explicit constructor calls, initialization list, or implicitly in a member/variable declaration.
+A variable in C++ can also be in an uninitialized state after its declaration. HLSL inherits most of these behvior from C++ by allowing variables to be uninitialized.
+
+On the other hand, languages like C# and Swift has a set of well defined rules to ensure every variable is initialized after its declaration.
+C++ allows using the initilization list syntax to initialize an object. The semantics of initialization lists depends on whether or not explicit constructors
+are defined on the type.
+
+Proposed Approach
+-----------------
+
+In this section, we document all concepts and rules related to initialization, constructors and initialization lists.
+
+### Default Initializable type
+
+A type is considered "default-initializable" if it provides a constructor that can take 0 arguments, so that it can be constructed with `T()`.
+
+### Variable Initialization
+
+Generally, a variable is considered uninitialized at its declaration site without an explicit value expression.
+For example,
+```csharp
+void foo()
+{
+  MyType t; // t is considered uninitialized.
+  var t1 : MyType; // same in modern syntax, t1 is uninitialized.
+}
+```
+
+However, the Slang language has been allowing implicit initialization of variables whose types are default initializable types.
+For example,
+```csharp
+struct MyType1 {
+  int x;
+  __init() { x = 0; }
+}
+void foo() {
+  MyType t1; // `t1` is initialized with a call to `__init`.
+}
```

Comment:
can you also make an example that if there is no `__init()` defined for `MyType1`, what will happen?